XMonad/Layout/IndependentScreens.hs

```diff
@@ -186,3 +218,86 @@ unmarshallWindowSpace :: WindowSpace -> WindowSpace
 
 marshallWindowSpace s ws = ws { tag = marshall s  (tag ws) }
 unmarshallWindowSpace ws = ws { tag = unmarshallW (tag ws) }
+
+-- for marshallPP to work with XMonad.Actions.CopyWindow
+-- only ppHidden is changed
+marshallPP' :: ScreenId -> PP -> PP
+marshallPP' s pp = pp {
+    ppCurrent           = ppCurrent         pp . snd . unmarshall,
+    ppVisible           = ppVisible         pp . snd . unmarshall,
+    ppHidden            = ppHidden          pp ,
+    ppHiddenNoWindows   = ppHiddenNoWindows pp . snd . unmarshall,
+    ppUrgent            = ppUrgent          pp . snd . unmarshall,
+    ppSort              = fmap (marshallSort' s) (ppSort pp)
+    }
+
+-- if use original marshallSort, getSortByIndex will not work properly, since it sends the "video" instead of "0_video" back, and getSortByIndex has no way of knowing the order of "video" if the workspaces are declared with withScreens. However, getSortByTag works. 
+marshallSort' :: ScreenId -> ([WindowSpace] -> [WindowSpace]) -> ([WindowSpace] -> [WindowSpace])
+marshallSort' s vSort = pScreens . vSort . vScreens where
+        onScreen ws = unmarshallS (tag ws) == s
+        vScreens    = filter onScreen
+        pScreens    = id
+
+-- do something with the Nth workspace
+-- can be used with >9 workspaces
+-- to let withNthWorkspace in XMonad.Actions.DynamicWorkspaces be screen aware
+withNthWorkspace' :: (String -> WindowSet -> WindowSet) -> Int -> X ()
+withNthWorkspace' job wnum = do 
+                                sort <- getSortByIndex
+                                ws <- gets (sort . XMonad.StackSet.workspaces . windowset)
+                                sc <- gets (screen . current . windowset)
+                                case test wnum ws sc of
+                                     (w:_) -> windows $ job w
+                                     [] -> return ()
+                                     where test wnum ws sc = drop wnum [ tag j | j<-ws, wsConvert j == sc]
+
+-- screen aware wrapper for copy from XMonad.Actions.CopyWindow
+copyWinTo :: Int->X ()
+copyWinTo j = do
+            sort <- getSortByIndex
+            ws <- gets (sort . XMonad.StackSet.workspaces . windowset)
+            sc <- gets (screen . current . windowset)
+            let wsID = [ tag m | m<-ws, wsConvert m == sc]
+            windows $ copy $ wsID !! (j-1)
+
+-- return current and visible workspaces on a specific monitor
+lookupWorkspace' :: (Eq s)=> s -> StackSet i l a s sd-> [i]
+lookupWorkspace' sc w = [ tag j | Screen j s _ <- current w : visible w , s == sc ] 
+
+-- take a list of Strings such as ["1","2","3","4"] -> 1234
+fromDigits :: [Char] -> Int
+fromDigits = foldl addDigit 0
+   where addDigit num d = 10*num + ord d - 48 
```

Comment:
I this guaranteed to always succeed?  If yes, then perhaps `read . head . sequence` would be clearer.  If not, then it's best to return a `Maybe Int` in either case (for the second option you can use `readMaybe`, for your option you'd need to see if `ord d` is actually a digit).

Reply:
So I have deleted it from the latest commit.